landlab/graph/framed_voronoi/framed_voronoi.py

```diff
@@ -0,0 +1,483 @@
+from ..graph import Graph
+from ..voronoi.voronoi import DelaunayGraph
+
+
+class HorizontalRectVoronoiGraph:
+    """This static class implements the horizontal rectangular frame for
+    the FramedVoronoiGraph.
+    """
+
+    @staticmethod
+    def number_of_nodes(shape):
+        """
+        Parameters
+        ----------
+        shape: tuple of int
+            Number of rows and number of columns
+
+        Returns
+        -------
+        Int
+            Number of nodes
+
+        Examples
+        --------
+        >>> from landlab.graph.framed_voronoi.framed_voronoi import HorizontalRectVoronoiGraph
+        >>> HorizontalRectVoronoiGraph.number_of_nodes((3, 2))
+        6
+        """
+        n_rows, n_cols = shape
+        return n_rows * n_cols
+
+    @staticmethod
+    def xy_of_node(
+        shape,
+        xy_spacing=(1.0, 1.0),
+        xy_of_lower_left=(0.0, 0.0),
+        xy_min_spacing=(0.5, 0.5),
+        random_seed=True,
+        seed=(200, 500),
+    ):
+        """Calculation of the x-y coordinates this way:
+        1. Generate a rectangular, regular meshgrid.
+        2. Move the coordinates of the core nodes over a random distance around their
+        ....initial position, within a threshold calculated from xy_spacing and
+        ....xy_min_spacing.
+        3. Rectify the y-coordinates of the nodes of the left and right to ensure
+        ....that the leftmost node of a row has a lower y than the rightmost node.
+        ....This ensures that the ids of these nodes are not modified by subsequent
+        ....sorting operations on the graph and make it possible to get the
+        ....perimeter nodes in simple way.
+
+        Parameters
+        ----------
+        shape : int or tuple of int.
+            For a rectangular layout, number of rows and columns of nodes.
+            If int, rows number = columns number = value.
+        xy_spacing : float or tuple of float, optional.
+            Node spacing along x and y coordinates. If float, same spacing at x and y.
+        xy_of_lower_left : tuple, optional.
+            Minimum x-of-node and y-of-node values. Depending on the grid.
+            no node may be present at this coordinate. Default is (0., 0.).
+        xy_min_spacing: float or tuple of float, optional.
+            Final minimal spacing between nodes. Random moves of the core nodes
+            around their position cannot be above this threshold:
+            (xy_spacing - xy_min_spacing) /2
+            If float, same minimal spacing for x and y.
+        random_seed: bool, optional.
+            If True, the moves of coordinates are completely random. False is used
+            when reproducibility of moves is needed. Move pseudo-randomness is
+            then controlled by the parameter seed.
+        seed: int or tuple of int, optional.
+            Seeds used to generate the random x and y moves. This parameter is unused
+            when random_seed = True.
```

Comment:
As written, *seed* must be a *tuple* not an *int*. It seems to me though that *seed* should only be allowed to be an *int*. I don't see the usefulness of having a separate seed for the *x* and *y* moves.

Reply:
i simplified by requiring a tuple of int only (or None).

x moves should be independent from y moves. 

Implementation pull a random move depending on the node id. If only one seed is used, this may imply to pull a random move on the node id for x and on the node id + number of nodes for y, which is less clearer in my view than just have independent seeds for x and y and make the pull on the node id similarly for the x and y move

(sorry if I expressed it in a not so clear way :flushed:)

Reply:
@sebastien-lenard Sorry, I still don't get why a user would want to have separate seeds for the *x* and *y* moves. The only reason I can think of is if they wanted the moves to be the same (i.e. *not* independent).

How about, instead, something like the following (*seed* is either an `int` or `None`),
```python
rng = np.random.default_rng(seed=seed)
x_moves = rng.uniform(-max_move[0], max_move[0], shape)
y_moves = rng.uniform(-max_move[1], max_move[1], shape)

x_of_node[1:-1, 1:-1] += x_moves[1:-1, 1:-1]
y_of_node[1:-1, 1:-1] += y_moves[1:-1, 1:-1]
```